### PR TITLE
Refactor Experiments code generation

### PR DESCRIPTION
**Pull request by patxibocos**
- Prefixed DTO classes that are mapping to Optimizely responses with `Optimizely`
- Prefixed base generator classes with `Base`
- Changed the Experiments generator to build experiment objects and its variants + client in a single function:

```kotlin
                optimizelyExperiments.forEach { experiment ->
                    val variantsEnumClassName = ClassName(packageName, "${experiment.key}Variants")
                    addType(experimentVariantsEnumTypeSpec(variantsEnumClassName, experiment.variations))
                    addType(experimentObjectTypeSpec(variantsEnumClassName, experiment))
                }
                addType(buildExperimentsClient(optimizelyExperiments))
```

### CODE REVIEW

**Comment by patxibocos** [comment]
I am also planning to remove the abstract class for the Experiments generator, what do you think @ralphpina?

File: src/test/kotlin/com/patxi/poetimizely/generator/ExperimentsGeneratorTest.kt [hunk 1]
```diff
@@ -14,53 +16,43 @@ import io.kotest.matchers.collections.shouldHaveSize
...
-            sources = kotlinSource
-            inheritClassPath = true
-            messageOutputStream = System.out
-        }.compile()
-    }
+class ExperimentsGeneratorTest : BehaviorSpec({
 
     given("An Optimizely experiment") {
         val experimentKey = "TEST_EXPERIMENT"
         val variationKey = "TEST_VARIATION"
```
Comment (ralphpina) [inline src/test/kotlin/com/patxi/poetimizely/generator/ExperimentsGeneratorTest.kt]: If you change these to "TEST-EXPERIMENT" the compilation fails:
```
v: Using Kotlin home directory <no_path>
v: Using JDK home directory /Library/Java/JavaVirtualMachines/jdk1.8.0_201.jdk/Contents/Home
v: Scripting plugin will not be loaded: not all required jars are present in the classpath (missing files: [./kotlin-scripting-compiler.jar, ./kotlin-scripting-compiler-impl.jar, ./kotlin-scripting-common.jar, ./kotlin-scripting-jvm.jar, ./kotlin-scripting-js.jar, ./js.engines.jar])
v: Configuring the compilation environment
e: /var/folders/39/zkb6v8qn3sg3z4nd_bln3qtw0000gn/T/Kotlin-Compilation7615595497319236156/sources/Experiments.kt: (27, 45): Unresolved reference: TEST
e: /var/folders/39/zkb6v8qn3sg3z4nd_bln3qtw0000gn/T/Kotlin-Compilation7615595497319236156/sources/Experiments.kt: (27, 50): Unresolved reference: EXPERIMENT

expected: OK but was: COMPILATION_ERROR
org.opentest4j.AssertionFailedError: expected: OK but was: COMPILATION_ERROR
	at com.patxi.poetimizely.generator.ExperimentsGeneratorTest$1$1$1$1$1.invokeSuspend(ExperimentsGeneratorTest.kt:38)
	at com.patxi.poetimizely.generator.ExperimentsGeneratorTest$1$1$1$1$1.invoke(ExperimentsGeneratorTest.kt)
	at io.kotest.core.spec.style.WhenContext$addThenContext$2.invokeSuspend(behaviorSpecDsl.kt:154)
	at io.kotest.core.spec.style.WhenContext$addThenContext$2.invoke(behaviorSpecDsl.kt)
```

Does it appear that some compiler is splitting the name and adding two files?

Is "TEST-EXPERIMENT" a valid experiment id? 

I had a similar use case, to map the feature key to something more "natural" for an enum constant with `[optimizelyFeatureKeyToEnumConstant](https://github.com/patxibocos/poetimizely/blob/master/src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt#L41)()`.

File: src/main/kotlin/com/patxi/poetimizely/generator/ExperimentsGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import com.squareup.kotlinpoet.TypeSpec
+import java.io.StringWriter
+
+class ExperimentsGenerator(private val packageName: String = "") {
+
+    fun build(optimizelyExperiments: List<OptimizelyExperiment>): String =
+        FileSpec.builder(packageName, "Experiments")
+            .apply {
+                optimizelyExperiments.forEach { experiment ->
+                    val variantsEnumClassName = ClassName(packageName, "${experiment.key}Variants")
```
Comment (ralphpina) [inline src/main/kotlin/com/patxi/poetimizely/generator/ExperimentsGenerator.kt]: This code would produce enums with names that don't look natural like `TEST_EXPERIMENTVariants`.

Maybe we generate `TestExperimentsVariants` from keys with names like `TEST_EXPERIMENTS` or `test-experiments`.

File: src/main/kotlin/com/patxi/poetimizely/generator/ExperimentsGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    private fun experimentVariantsEnumTypeSpec(
+        variantsEnumClassName: ClassName,
+        optimizelyVariations: Collection<OptimizelyVariation>
+    ): TypeSpec =
+        TypeSpec.enumBuilder(variantsEnumClassName)
+            .addSuperinterface(BaseVariant::class).apply {
+                optimizelyVariations.forEach { variation ->
+                    addEnumConstant(
+                        variation.key,
```
Comment (ralphpina) [inline src/main/kotlin/com/patxi/poetimizely/generator/ExperimentsGenerator.kt]: Should we parse these ids? For features there are tight restrictions on how an ID can be formated. Do we have similar guarantees for experiments? If so maybe we can convert something like `test-experiment` to `TEST_EXPERIMENT`?

File: src/main/kotlin/com/patxi/poetimizely/generator/ExperimentsGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                        ).build()
+                    ).build()
+                }
+            }.build()
+
+    private fun experimentObjectTypeSpec(
+        variantsEnumClassName: ClassName,
+        optimizelyExperiment: OptimizelyExperiment
+    ): TypeSpec =
+        TypeSpec.objectBuilder(ClassName(packageName, optimizelyExperiment.key)).apply {
```
Comment (ralphpina) [inline src/main/kotlin/com/patxi/poetimizely/generator/ExperimentsGenerator.kt]: Should we convert a key to a more "natural" Kotlin class name such as `TestExperiment`?

File: src/main/kotlin/com/patxi/poetimizely/generator/ExperimentsGenerator.kt [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        TypeSpec.classBuilder(ClassName(packageName, "ExperimentsClient"))
+            .primaryConstructor(
+                FunSpec.constructorBuilder().addParameter("optimizely", Optimizely::class)
+                    .addParameter("userId", String::class).build()
+            )
+            .superclass(BaseExperimentsClient::class)
+            .addSuperclassConstructorParameter("optimizely")
+            .addSuperclassConstructorParameter("userId")
+            .addFunction(
+                FunSpec.builder("getAllExperiments").addModifiers(KModifier.OVERRIDE)
```
Comment (ralphpina) [inline src/main/kotlin/com/patxi/poetimizely/generator/ExperimentsGenerator.kt]: nice, I thought you needed to specify the return type 👍.

File: src/test/kotlin/com/patxi/poetimizely/generator/ExperimentsGeneratorTest.kt (hunk 1 above)
Comment (patxibocos) [reply src/test/kotlin/com/patxi/poetimizely/generator/ExperimentsGeneratorTest.kt]: I am going to follow the same approach as those restrictions also apply for experiments and variations :)

**Comment by patxibocos** [comment]
I just made the changes but in #17 by mistake 🤦‍♂ https://github.com/patxibocos/poetimizely/pull/17/commits/fc9ee9770426cd2087e6d4e39be2c4e6e1bae2dd
Do you mind reviewing that again and if that is ok we can just merge that and then merge this finally to master?

**Comment by ralphpina** [comment]
👍 approved the other. It looks great 🎉.

**Pull request merged**